runner/query_run.go

```diff
@@ -0,0 +1,131 @@
+package runner
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pganalyze/collector/input/postgres"
+	"github.com/pganalyze/collector/output/pganalyze_collector"
+	"github.com/pganalyze/collector/state"
+	"github.com/pganalyze/collector/util"
+)
+
+func SetupQueryRunnerForAllServers(ctx context.Context, servers []*state.Server, collectionOpts state.CollectionOpts, logger *util.Logger) {
+	if collectionOpts.ForceEmptyGrant {
+		return
+	}
+	for idx := range servers {
+		go func(server *state.Server) {
+			logger = logger.WithPrefixAndRememberErrors(server.Config.SectionName)
+			cleanupInterval := time.NewTicker(5 * time.Minute)
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-cleanupInterval.C:
+					cleanup(server)
+				default:
+					if server.Config.EnableQueryRunner {
+						run(ctx, server, collectionOpts, logger)
+					}
+					time.Sleep(1 * time.Second)
+				}
+			}
+		}(servers[idx])
+	}
+}
+
+func run(ctx context.Context, server *state.Server, collectionOpts state.CollectionOpts, logger *util.Logger) {
+	for idx, query := range server.QueryRuns {
+		if !query.FinishedAt.IsZero() {
+			continue
+		}
+		server.QueryRunsMutex.Lock()
+		server.QueryRuns[idx].StartedAt = time.Now()
+		server.QueryRunsMutex.Unlock()
+		logger.PrintVerbose("Query run %d starting: %s", query.Id, query.QueryText)
+
+		db, err := postgres.EstablishConnection(ctx, server, logger, collectionOpts, query.DatabaseName)
+		if err != nil {
+			server.QueryRunsMutex.Lock()
+			server.QueryRuns[idx].FinishedAt = time.Now()
+			server.QueryRuns[idx].Error = err.Error()
+			server.QueryRunsMutex.Unlock()
+			continue
+		}
+		defer db.Close()
+
+		err = postgres.SetStatementTimeout(ctx, db, 60*1000)
+		if err != nil {
+			server.QueryRunsMutex.Lock()
+			server.QueryRuns[idx].FinishedAt = time.Now()
+			server.QueryRuns[idx].Error = err.Error()
+			server.QueryRunsMutex.Unlock()
+			continue
+		}
+
+		pid := 0
+		err = db.QueryRow(postgres.QueryMarkerSQL + "SELECT pg_backend_pid()").Scan(&pid)
+		if err == nil {
+			server.QueryRunsMutex.Lock()
+			server.QueryRuns[idx].BackendPid = pid
+			server.QueryRunsMutex.Unlock()
+		} else {
+			server.QueryRunsMutex.Lock()
+			server.QueryRuns[idx].FinishedAt = time.Now()
+			server.QueryRuns[idx].Error = err.Error()
+			server.QueryRunsMutex.Unlock()
+			continue
+		}
+
+		// We don't include QueryMarkerSQL so query runs are reported separately in pganalyze
+		comment := fmt.Sprintf("/* pganalyze:no-alert,pganalyze-query-run:%d */ ", query.Id)
+		prefix := ""
+		result := ""
+		if query.Type == pganalyze_collector.QueryRunType_EXPLAIN {
+			prefix = "EXPLAIN (ANALYZE, VERBOSE, BUFFERS, FORMAT JSON) "
+		}
+
+		err = db.QueryRowContext(ctx, comment+prefix+query.QueryText).Scan(&result)
```

Comment:
In today's meeting, we talked about wrapping this in the transaction (maybe the below is not needed as it's a simple EXPLAIN - or maybe we could wrap this in the transaction even wider).

Also I completely forgot about this, but I think we actually wanted to run this 3 times, then record the 3rd one (or fastest one), and we don't have that logic in place in this PR.

Reply:
Why would we want to run it three times? It should already be a warm cache after two runs (which this PR already does).

Reply:
Ah 🤦 I wasn't looking at the line below this one, sorry, my bad.
So this is the first run, and line 96 is second, then there is the 3rd one in line 100 (and the explain without analyze is in line 106). Please disregard what I said re: running 3 times.